draft-ietf-httpbis-priority.md

```diff
@@ -634,6 +634,32 @@ signal the intended priority to the client by including the Priority field in
 a PUSH_PROMISE or HEADERS frame.
 
 
+# Retransmission Scheduling
+
+A transport may detect packet losses and then retranmit lost information to
+provide reliability. Examples include TCP and QUIC. While this document
+specifies HTTP layer prioritization, its effectiveness can be further achieved
+if transport layer also respect such priorities. Besides scheduling new data,
+transport implementations MAY also consider stream priorities during
+retransmission. In this section we will use QUIC as an example to discusss a
+couple scenarios an implementation need to consider.
+
+{{!I-D.ietf-quic-transport}}, Section 13.3 states that endpoints SHOULD 
+prioritize retransmission over sending new data, unless otherwise specified by
+application. When an HTTP/3 application uses priority scheme defined in this 
+document, the QUIC transport layer MAY take stream priorities as input to 
+scheduling the sending of both retransmission data and new data. For example, 
+if one stream with higher urgency has new data to send and another stream with 
+lower urgency has retransmission data to send, a QUIC transport may choose to 
+schedule the higher urgency new data before the lower urgency retransmission
+data.
```

Comment:
```suggestion
{{!I-D.ietf-quic-transport}}, Section 13.3 states "Endpoints SHOULD prioritize
retransmission of data over sending new data, unless priorities specified by the
application indicate otherwise". When an HTTP/3 application uses the priority
scheme defined in this document, the QUIC transport implementation SHOULD
consider the relative priority of streams when scheduling of new data and
retransmission data. For example, if one stream with higher urgency has new data
to send and another stream with lower urgency has retransmission data to send, a
QUIC transport is more likely to fulfil the application's intent if it chooses
to schedule the higher urgency new data before the lower urgency retransmission
data. While there is no requirement on QUIC implementations to implement the
extensible priority scheme, an aligned prioritization model can help
applications manage scheduling expectations.
```

Reply:
I'm not sure we have enough data at this point to know if it should be a SHOULD or a MAY... AFAIK nobody has really researched the impact of this. IIUC FB implemented this, but hasn't looked too closely at metric impact yet (though I assume it's not massive in either direction or it would have been spotted). 

I'm mainly thinking about (theoretical) situations where there's just a single packet lost of a low prio stream and that's the only outstanding data for that stream, where it might make sense to retrans that packet before higher prio data etc. (you can extrapolate that to X packets, where X is a certain percentage of estimated cwnd etc.). Similar remarks can be made for potential starvation, app-layer logic that might abort streams if they take too long to complete, etc. 

I would say it's probably best to go with the conservative MAY for now, mainly to make sure people are aware of this possibility.

Reply:
Thanks Robin. 

Since we don't have much info I agree it's hard to make any recommendation. We should be wary of suggsting something that turns out to be bad.

 To paraphrase the point I was trying to communicate, QUIC transport says "SHOULD prioritize restransmissions unless there's some communicated application-intent" and here we are saying "if the transport layer speaks a compatible model, it really SHOULD factor in stream priorities, but what chooses to do is still up to it". 

If you agree with the premise of that point, I think I can highlight it better by changing words around a bit such as

> 
            When an HTTP/3 application uses the priority
            scheme defined in this document and the QUIC transport implementation
            supports application indicated stream priority, a transport that considers
            the relative priority of streams when scheduling both new data and
            retransmission data might better match the expectations of the application.
            However, here are no requirements on how it chooses to schedule based on
            this information because the decision depends on several factors and trade-offs,
            A transport could prioritize new data for a higher urgency stream over retransmission data
            for a lower priority stream, or it could prioritize retransmission data over new data 
            irrespective of urgencies.

Reply:
I think this newer text indeed better reflects what we should be saying here, and also highlights the tradeoffs.